### ISSUE TITLE
Memory leaks from renderPlot

### ISSUE THREAD

**Issue by wch** [rstudio/shiny#2423]
This test app shows various ways `renderPlot()` and assigning to outputs can leak memory. (This is originally from https://community.rstudio.com/t/memory-leak-when-using-ggplot2-grobs-and-tabs-shiny-or-my-bug/30108/.) **Update:** I've also added examples with `renderText()`.

In the first case, `renderPlot()` is called in an observer and assigned to an output each time. It leaks about 76528 bytes each time.

In the second case, a `renderPlot()` is invalidated. This leaks 2352 bytes each time.

In the third case, `renderPlot()` is called in an observer, but it is simply discarded. This leaks 1480 bytes each time.

The next three cases are the same, but use `renderText()`. They leak 2352, 1168, and 1168 bytes, respectively.

```R
library(shiny)
library(pryr)

ui <- fluidPage(
  sidebarLayout(
    sidebarPanel(
      div(
        "In observer, call renderPlot() and assign to output", br(),
        actionButton(inputId = "replot_1", label = "Go")
      ),
      br(),
      div(
        "Invalidate renderPlot()", br(),
        actionButton(inputId = "replot_2", label = "Go" )
      ),
      br(),
      div(
        "In observer, call renderPlot() but don't assign to an output", br(),
        actionButton(inputId = "replot_3", label = "Go" )
      ),
      br(),
      div(
        "In observer, call renderText() and assign to output", br(),
        actionButton(inputId = "retext_1", label = "Go")
      ),
      br(),
      div(
        "Invalidate renderText()", br(),
        actionButton(inputId = "retext_2", label = "Go" )
      ),
      br(),
      div(
        "In observer, call renderText() but don't assign to an output", br(),
        actionButton(inputId = "retext_3", label = "Go" )
      ),
      br(),
      verbatimTextOutput("memory_info")
    ),
    mainPanel(
      plotOutput("plot_1"),
      plotOutput("plot_2"),
      textOutput("text_1"),
      textOutput("text_2")
    )
  )
)

server <- function(input, output, session) {
  observeEvent(input$replot_1, {
    output$plot_1 <- renderPlot({
      plot(1:20, 1:20)
    })
  })

  output$plot_2 <- renderPlot({
    input$replot_2
    plot(1:20, 1:20)
  })

  observeEvent(input$replot_3, {
    renderPlot({
      plot(1:20, 1:20)
    })
  })

  observeEvent(input$retext_1, {
    output$text_1 <- renderText({
      "Text 1"
    })
  })

  output$text_2 <- renderText({
    input$retext_2
      "Text 2"
  })

  observeEvent(input$retext_3, {
    renderText({
      "Text 3"
    })
  })


  last_mem_used <- 0L
  output$memory_info <- renderText({
    input$replot_1
    input$replot_2
    input$replot_3
    input$retext_1
    input$retext_2
    input$retext_3
    gc()
    current_mem_used <- mem_used()
    res <- sprintf(
      "mem_used(): %d bytes\nChange: %d bytes",
      current_mem_used,
      current_mem_used - last_mem_used
    )
    last_mem_used <<- current_mem_used

    res
  })
}

shinyApp(ui = ui, server = server)
```

**Comment by wch** [comment]
After some investigation with @jcheng5, we've found the following: Some of the smaller memory leaks are due to the R symbol leakage issue, and these are solved by #2429. However, the large memory leak that happens with `output$plot_1` has a different cause.

`renderPlot()` is a bit different from other render functions. All of the `render*` functions return a function, but the one returned by `renderPlot()` makes use of a reactive expression. When a the function is saved as an output, the [`defineOutput`](https://github.com/rstudio/shiny/blob/64b3095f2cfa96d7ccc513a82de58edb14bb91ed/R/shiny.R#L1064) method wraps it in an observer. When that output is replaced with another one, there's nothing that will consume the previous reactive expression. However, the reactive expression will not be garbage collected, because there are some reactive values that it used that still refer to it, namely, the width, height, and pixelratio from `clientData`. If there is something that invalidates the reactive expression, then those references will be discarded and the reactive expression can be GC'd.

Here is a reactlog diagram that illustrates. The observer has been invalidated, but the `plotObj` reactive expression is still hanging around because of the incoming arrows.

![image](https://user-images.githubusercontent.com/86978/57551286-46801600-732e-11e9-85c4-423a0880473e.png)


This example app shows one way to explicitly invalidate the reactive expression. The second plot uses `input$replot_2`, and it leaks much less memory than the first plot.

```R
ui <- fluidPage(
  sidebarLayout(
    sidebarPanel(
      div(
        "In observer, call renderPlot() and assign to output", br(),
        actionButton(inputId = "replot_1", label = "Go")
      ),
      br(),
      div(
        "In observer, call renderPlot() and assign to output. The reactive expression itself takes a dependency on input$replot_2",
        br(),
        actionButton(inputId = "replot_2", label = "Go")
      ),
      br(),
      verbatimTextOutput("memory_info")
    ),
    mainPanel(
      plotOutput("plot_1"),
      plotOutput("plot_2")
    )
  )
)

server <- function(input, output, session) {
  observeEvent(input$replot_1, {
    output$plot_1 <- renderPlot({
      ggplot(mtcars, aes(wt, mpg)) + geom_point()
    })
  })

  observeEvent(input$replot_2, {
    output$plot_2 <- renderPlot({
      input$replot_2
      ggplot(mtcars, aes(wt, mpg)) + geom_point()
    })
  })

  last_mem_used <- 0L
  output$memory_info <- renderText({
    input$replot_1
    input$replot_2
    gc()
    current_mem_used <- mem_used()
    res <- sprintf(
      "mem_used(): %d bytes\nChange: %d bytes",
      current_mem_used,
      current_mem_used - last_mem_used
    )
    last_mem_used <<- current_mem_used

    message(res)
    res
  })
}

shinyApp(ui = ui, server = server)
```


We have two potential ways of solving this. The first is more expedient: We could modify `markRenderFunc()` so that it can accept a cleanup callback. For `renderPlot()`, the cleanup callback would explicitly invalidate the reactive expression. We would also modify the `Observable` class so that it has an `invalidate()` method.

The second potential solution is to implement and use weak references. The arrows in the reactive graph would be weak references, so they would not, on their own, prevent the reactive expression from being garbage collected.

**Comment by wch** [comment]
Here's an example that shows the problem more directly. (Note that the memory values are after running this a few times, because R does some stuff that increases memory usage the first few times, but then it stops.)

```R
library(shiny)
library(pryr)
rm(list=ls())

mem_used()
#> 49 MB

v <- reactiveVal(1)

r <- reactive({
  v()
  # Use sample() to avoid ALTREP optimization of 1:1e6
  sample(1:1e6)
})

o <- observe({
  r()
})

mem_used()
#> 49 MB

shiny:::flushReact()
mem_used()
#> 53 MB

rm(r, o)
mem_used()
#> 53 MB

rm(v)
mem_used()
#> 49 MB
```

Even after `r` and `o` are removed, `r` can't be garbage collected, because `v` still has a reference to it (so that `v` can invalidate `r` in the future if `v` changes).

Another way to release `r` is, instead of calling `rm(v)`, to invalidate `r` by changing `v()`. Replacing `rm(v)` with `v(2)` gives the following:

```R
v(2)
mem_used()
#> 49 MB
```

In the situation where the user writes `r`, but is given `v` (and doesn't have control over it), another solution is to have `r` take a dependency on another reactive value that the user controls. In this example, `cleanup` is used for that purpose.

```R
library(shiny)
library(pryr)
rm(list=ls())

mem_used()
#> 49 MB

# User doesn't have control over v
v <- reactiveVal(1)
# User does have control over cleanup
cleanup <- reactiveVal(1)

r <- reactive({
  v()
  cleanup()
  # Use sample() to avoid ALTREP optimization of 1:1e6
  sample(1:1e6)
})
o <- observe({
  r()
})

mem_used()
#> 49.2 MB

shiny:::flushReact()
mem_used()
#> 53.2 MB

rm(r, o)
mem_used()
#> 53.2 MB

cleanup(2)
mem_used()
#> 49.2 MB
```
